### PR TITLE
Debugger

### PR DESCRIPTION
**Pull request by ghallak**

### CODE REVIEW

File: src/aere_utils.erl [hunk 1]
```diff
@@ -0,0 +1,46 @@
+-module(aere_utils).
```
Comment (radrow) [inline src/aere_utils.erl:1]: `aere_files`? It only does file stuff, no need to call it utils I think?

**Review by radrow — APPROVED** [review]

**Review by radrow — CHANGES_REQUESTED** [review]
Found some crash

**Comment by radrow** [comment]
Crash:
```
AESO> :l C.aes
AESO> let c = Chain.create() : C
Command failed:
{badmatch,#{}}
Stacktrace:
[{aeso_ast_infer_types,ets_tabid,1,
                       [{file,"/home/rjro/Documents/a/aerepl/_build/default/lib/aesophia/src/aeso_ast_infer_types.erl"},
                        {line,2492}]},
 {aeso_ast_infer_types,ets_insert,2,
                       [{file,"/home/rjro/Documents/a/aerepl/_build/default/lib/aesophia/src/aeso_ast_infer_types.erl"},
                        {line,2513}]},
 {aeso_ast_infer_types,lookup_env1,4,
                       [{file,"/home/rjro/Documents/a/aerepl/_build/default/lib/aesophia/src/aeso_ast_infer_types.erl"},
                        {line,496}]},
 {aere_sophia,type_of_user_input,1,
              [{file,"/home/rjro/Documents/a/aerepl/src/aere_sophia.erl"},
               {line,65}]},
 {aere_repl,register_letval,3,
            [{file,"/home/rjro/Documents/a/aerepl/src/aere_repl.erl"},
             {line,208}]},
 {aere_gen_server,handle_call,3,
                  [{file,"/home/rjro/Documents/a/aerepl/src/aere_gen_server.erl"},
                   {line,66}]},
 {gen_server,try_handle_call,4,[{file,"gen_server.erl"},{line,1149}]},
 {gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,1178}]}]
```
`C.aes`:
```
contract C =
  entrypoint f(x) = x + 1
```

File: src/aere_parse.erl [hunk 2]
```diff
@@ -1,66 +1,72 @@
...
+       {[l], [{some, string}], "FILENAMES",
         [ "Loads files into the REPL. Adds the last file to the scope."
         , "Cleans user variables, functions, types and contracts."
         , "Unloads previously loaded files."]
        }}
-    , {"reload",
-       {["r"], many_args, "[FILENAMES]",
-        [ "Reloads given files preserving the included scope."
+    , {reload,
+       {["r"], [], "",
```
Comment (radrow) [inline src/aere_parse.erl:49]:
```suggestion
       {[r], [], "",
```

**Comment by radrow** [comment]
Crash:
```
AESO> :h
=ERROR REPORT==== 25-Apr-2023::23:58:31.950811 ===
** Generic server aere_gen_server terminating 
** Last message in was help
** When Server state == {rs,
                         {ready,
                          {state,
                           {state,
                            {trees,
                             {mpt,

...
** Reason for termination ==
** {badarg,[{erlang,'++',
                    [[10,45,32,58|stacktrace],
                     "\nType `:help COMMAND` to learn about the given command"],
                    [{error_info,#{module => erl_erts_errors}}]},
            {lists,append,1,[{file,"lists.erl"},{line,155}]},
            {lists,append,1,[{file,"lists.erl"},{line,155}]},
            {string,join,2,[{file,"string.erl"},{line,2244}]},
            {aere_msg,help,0,
```

File: src/aere_sophia.erl [hunk 3]
```diff
@@ -28,12 +53,12 @@ typecheck(Ast, Opts) ->
...
-        {Code, _} -> Code
+    Opts = [debug_info, include_child_contract_symbols | optimizations_off()],
+    {#{child_con_env := ChildConEnv, saved_fresh_names := SavedFreshNames}, FCode}
+        = try aeso_ast_to_fcode:ast_to_fcode(TypedAst, Opts)
+          catch {error, Ec} -> process_err(Ec) end,
+    try
+        aeso_fcode_to_fate:compile(ChildConEnv, FCode, SavedFreshNames, Opts)
     catch {error, Ef} -> process_err(Ef) end.
 
 type_of_user_input(TEnv) ->
```
Comment (radrow) [inline src/aere_sophia.erl:64]: Crash:
```
AESO> 2
Command failed:
{badmatch,#{}}
Stacktrace:
[{aeso_ast_infer_types,ets_tabid,1,
....
```
I am using OTP 25, but I don't think it's an issue. It looks like a bug in the compiler actually. Do all tests pass there?

File: src/aere_msg.erl [hunk 4]
```diff
@@ -254,7 +295,7 @@ help() ->
 
 -spec help(any()) -> msg().
 help(What) ->
```
Comment (radrow) [inline src/aere_msg.erl:297]: Crash:
```
AESO> :h
=ERROR REPORT==== 25-Apr-2023::23:58:31.950811 ===
** Generic server aere_gen_server terminating 
** Last message in was help
** When Server state == {rs,
                         {ready,
                          {state,
                           {state,
                            {trees,
                             {mpt,

...
** Reason for termination ==
** {badarg,[{erlang,'++',
                    [[10,45,32,58|stacktrace],
                     "\nType `:help COMMAND` to learn about the given command"],
                    [{error_info,#{module => erl_erts_errors}}]},
            {lists,append,1,[{file,"lists.erl"},{line,155}]},
            {lists,append,1,[{file,"lists.erl"},{line,155}]},
            {string,join,2,[{file,"string.erl"},{line,2244}]},
            {aere_msg,help,0,
```

File: src/aere_debugger.erl [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        , stacktrace/1
+        ]).
+
+
+-spec add_breakpoint(ReplState, FileName, Line) -> ReplState
+    when ReplState :: aere_repl_state:state(),
+         FileName  :: string(),
+         Line      :: integer().
+
+add_breakpoint(State, FileName, Line) ->
```
Comment (radrow) [inline src/aere_debugger.erl:17]: One can add a breakpoint in a file that does not exist. This is super prone to typos, I think we should restrict only to files that are loaded.

File: src/aere_sophia.erl (hunk 3 above)
Comment (radrow) [reply src/aere_sophia.erl:64]: Same for `"type`. Interestingly, it displays type errors correctly.

**Review by radrow — CHANGES_REQUESTED** [review]

File: src/aere_utils.erl (hunk 1 above)
Comment (ghallak) [reply src/aere_utils.erl]: Changed here 88f1c40eca14d7a9de43ef291f39b33b324c877f.

File: src/aere_debugger.erl (hunk 5 above)
Comment (ghallak) [reply src/aere_debugger.erl:17]: Done here b593859aaec13f07a04c21ce9459f8e022d3c894

File: src/aere_msg.erl (hunk 4 above)
Comment (ghallak) [reply src/aere_msg.erl:303]: Fixed here 4aeaf2f17ff91280fcbe99a379b7243426009662

File: src/aere_sophia.erl (hunk 3 above)
Comment (ghallak) [reply src/aere_sophia.erl]: Did I fix this here 0f1716f52973622f4001eb628c967814a44c365e? ;)

**Review by radrow — APPROVED** [review]

**Review by radrow — APPROVED** [review]

**Review by radrow — APPROVED** [review]

**Review by radrow — APPROVED** [review]

**Pull request merged**